python/ray/serve/_private/request_router/prefix_aware_router.py

```diff
@@ -0,0 +1,333 @@
+# These imports are used for metrics tracking, will remove for PR
+import logging
+import time
+from typing import (
+    List,
+    Optional,
+)
+
+import ray
+from ray.llm._internal.serve.request_router.prefix_aware.prefix_tree import (
+    PrefixTreeActor,
+)
+from ray.serve._private.common import ReplicaID
+from ray.serve._private.constants import (
+    SERVE_LOGGER_NAME,
+)
+from ray.serve._private.replica_result import ReplicaResult
+from ray.serve._private.request_router import (
+    PowerOfTwoChoicesRequestRouter,
+)
+from ray.serve._private.request_router.common import (
+    PendingRequest,
+)
+from ray.serve._private.request_router.replica_wrapper import (
+    RunningReplica,
+)
+from ray.serve._private.request_router.request_router import (
+    LocalityMixin,
+    MultiplexMixin,
+    RequestRouter,
+)
+
+logger = logging.getLogger(SERVE_LOGGER_NAME)
+
+
+class PrefixAwarePow2ReplicaRouter(LocalityMixin, MultiplexMixin, RequestRouter):
+    """Extends the PowerOfTwoChoicesRequestRouter with prefix-matching capabilities.
+
+    This request router optimizes replica selection by considering input text prefixes:
+
+    1. Mixes between three strategies to balance prefix cache hit rate and load balancing:
+       - When load is balanced (queue length difference < threshold), it selects replicas
+         with the highest prefix match rate for the input text
+       - When load is balanced but match rate is below 10%, it falls back to the smallest tenants
+       - When load is imbalanced, it uses the default Power of Two selection
+
+    2. Maintains a prefix tree to track which replicas have processed similar inputs:
+       - Inserts prompt text into the prefix tree after routing
+       - Uses this history to inform future routing decisions
+
+    This approach improves performance by routing related requests to the same replicas,
+    increasing cache locality and reducing overhead for language model inference.
+    """
+
+    def __init__(
+        self,
+        *args,
+        imbalanced_threshold=10,
```

Comment:
one thing the current api does not allow you to do is defining values for these parameters from the config perspective. We need to open this API up from serve @eicherseiji